### ISSUE TITLE
transition hooks - how to handle back button during asynchronous transition

### ISSUE THREAD

**Issue by dahei** [rackt/react-router#1237]
Hi,

we are litte confused how to handle the following use case:

- User clicks a link to another route
- the current component has a transition hook
    willTransitionFrom(transition, component, callback)

which is called.

Before calling the callback I want to animate some parts of the page. When the animation is done (e.g. after 2 seconds) I call the callback - it's all fine.

But, what is the best practice to handle an abort/back-button click during the animation respectively asynchronous flow *before* the callback is called.

Right now we used some kind of *hacky* solution:

    willTransitionFrom: function(transition, component, callback) {

        // normal flow - an async operation which takes 2 seconds
        // When done, the callback is called.
        doSomeAsyncStuff(callback);

       // HACK?
       // this function is called, when user clicks back button during the async operation
       //--> is this the only way to react to transition interception (e.g. back button was clicked) ? 
       transition.abort = function(reason) {
          // reason === "Cancellation" in the case of back-button use during async procedure
          // here we can roll back our animation for example and then call the callback
       }      
    }

Is this the way to go? The documentation and examples aren't clear enough to understand how it should be handled.

Thank you!

**Comment by ryanflorence** [comment]
I think 1.0 deals with this, @mjackson can you confirm?

**Comment by mjackson** [comment]
tl;dr: Don't use transition hooks for animation. Use [`TransitionGroup` and `CSSTransitionGroup`](https://facebook.github.io/react/docs/animation.html) instead.

Transition hooks are only for aborting or redirecting the transition. You should avoid doing async work in transition hooks in browsers because they make your UI non-responsive (the transition does not proceed until you call the callback). Async transition hooks can be useful when using the router server-side if you want to do things like persist state to a database.

**Comment by dahei** [comment]
The problem with CSSTransitionGroup is that it is limited to CSS-Transitions and animations. We use a lot GSAP and JS animations, because they are simply much more powerful.

We avoid the non-responsiveness  of the UI in aborting (and rewinding! if neccessary - possible with TimelineMax) the animation so the UI is always fully responsive and the url-state correct.

So maybe our take on this is interessting for other people using JS Animations.
